parser/proto.y

```diff
@@ -327,269 +334,231 @@ stringLit : _STRING_LIT {
+	| messageLiteralField ',' {
 		if $1 != nil {
-			$$ = ast.NewMessageFieldNode($1, $2, $3)
+			entry := &messageFieldEntry{$1, $2}
+                	$$ = &messageFieldList{entry, nil}
 		} else {
 			$$ = nil
 		}
 	}
-	| aggName '[' ']' {
+	| messageLiteralField ';' {
 		if $1 != nil {
-			val := ast.NewArrayLiteralNode($2, nil, nil, $3)
-			$$ = ast.NewMessageFieldNode($1, nil, val)
+			entry := &messageFieldEntry{$1, $2}
+			$$ = &messageFieldList{entry, nil}
 		} else {
 			$$ = nil
 		}
 	}
-	| aggName ':' '[' ']' {
+	| messageLiteralField {
 		if $1 != nil {
-			val := ast.NewArrayLiteralNode($3, nil, nil, $4)
-			$$ = ast.NewMessageFieldNode($1, $2, val)
+			entry := &messageFieldEntry{$1, nil}
+			$$ = &messageFieldList{entry, nil}
 		} else {
 			$$ = nil
 		}
 	}
-	| aggName '[' msgList ']' {
-		if $1 != nil {
-			vals, commas := $3.toNodes()
-			val := ast.NewArrayLiteralNode($2, vals, commas, $4)
-			$$ = ast.NewMessageFieldNode($1, nil, val)
-		} else {
-			$$ = nil
-		}
-	}
-	| aggName ':' '[' constantList ']' {
-		if $1 != nil {
-			vals, commas := $4.toNodes()
-			val := ast.NewArrayLiteralNode($3, vals, commas, $5)
-			$$ = ast.NewMessageFieldNode($1, $2, val)
-		} else {
-			$$ = nil
-		}
-	}
-	| aggName ':' '[' error ']' {
-		$$ = nil
-	}
-	| aggName '[' error ']' {
+	| {
 		$$ = nil
 	}
-	| aggName ':' msgLit {
-		if $1 != nil && $3 != nil {
+
+messageLiteralField : messageLiteralFieldName ':' value {
+		if $1 != nil {
 			$$ = ast.NewMessageFieldNode($1, $2, $3)
 		} else {
 			$$ = nil
 		}
 	}
-	| aggName msgLit {
-		if $1 != nil && $2 != nil {
+	| messageLiteralFieldName messageValue {
+		if $1 != nil {
 			$$ = ast.NewMessageFieldNode($1, nil, $2)
 		} else {
 			$$ = nil
 		}
 	}
-	| aggName ':' '<' error '>' {
+	| error ',' {
+		$$ = nil
+	}
+	| error ';' {
 		$$ = nil
 	}
-	| aggName '<' error '>' {
+	| error {
 		$$ = nil
 	}
 
-aggName : name {
+messageLiteralFieldName : identifier {
 		$$ = ast.NewFieldReferenceNode($1)
 	}
-	| '[' ident ']' {
+	| '[' qualifiedIdentifier ']' {
 		$$ = ast.NewExtensionFieldReferenceNode($1, $2.toIdentValueNode(nil), $3)
 	}
-	| '[' ident '/' ident ']' {
+	| '[' qualifiedIdentifier '/' qualifiedIdentifier ']' {
 		$$ = ast.NewAnyTypeReferenceNode($1, $2.toIdentValueNode(nil), $3, $4.toIdentValueNode(nil), $5)
 	}
 	| '[' error ']' {
 		$$ = nil
 	}
 
-msgList : msgLit {
-		if $1 == nil {
-			$$ = nil
-		} else {
-			$$ = &valueList{$1, nil, nil}
-		}
+value : scalarValue {
+		$$ = $1
 	}
-	| msgLit ',' msgList {
-		if $1 == nil {
-			$$ = nil
-		} else {
-			$$ = &valueList{$1, $2, $3}
-		}
+	| messageLiteral {
+		$$ = $1
+	}
+	| listLiteral {
+		$$ = $1
 	}
 
-msgLit : aggregate {
+messageValue : messageLiteral {
 		$$ = $1
 	}
-	| '<' aggFields '>' {
+	| listOfMessagesLiteral {
+		$$ = $1
+	}
+
+messageLiterals : messageLiteral ',' messageLiterals {
+		$$ = &valueList{$1, $2, $3}
+	}
+	| messageLiteral {
+		$$ = &valueList{$1, nil, nil}
+	}
```

Comment:
Move this below so it's next to where it's actually used, just before `listOfMessagesLiteral`.

Reply:
I moved this below.